src/transform/src/analysis/equivalences.rs

```diff
@@ -0,0 +1,560 @@
+                                expr1.clone().call_is_null().not(),
+                                expr2.clone().call_is_null().not(),
+                            ]);
+                        }
+                    }
+                    class.retain(|expr| {
+                        if let MirScalarExpr::CallBinary {
+                            func: mz_expr::BinaryFunc::Eq,
+                            expr1: _,
+                            expr2: _,
+                        } = expr
+                        {
+                            false
+                        } else {
+                            true
+                        }
+                    });
+                }
+            }
+            self.classes.extend(to_add);
+
+            // Tidy up classes, restore representative.
+            for class in self.classes.iter_mut() {
+                // TODO: Ideally we put literals as representatives, to reduce dependencies on rows.
+                // It is important that this order be "by complexity", so that
+                class.sort_by(|e1, e2| match (e1, e2) {
+                    (MirScalarExpr::Literal(_, _), MirScalarExpr::Literal(_, _)) => e1.cmp(e2),
+                    (MirScalarExpr::Literal(_, _), _) => std::cmp::Ordering::Less,
+                    (_, MirScalarExpr::Literal(_, _)) => std::cmp::Ordering::Greater,
+                    (MirScalarExpr::Column(_), MirScalarExpr::Column(_)) => e1.cmp(e2),
+                    (MirScalarExpr::Column(_), _) => std::cmp::Ordering::Less,
+                    (_, MirScalarExpr::Column(_)) => std::cmp::Ordering::Greater,
+                    (x, y) => {
+                        // General expressions should be ordered by their size,
+                        // to ensure we only simplify expressions by substititution.
+                        let x_size = x.size();
+                        let y_size = y.size();
+                        if x_size == y_size {
+                            x.cmp(y)
+                        } else {
+                            x_size.cmp(&y_size)
+                        }
+                    }
+                });
+                class.dedup();
+            }
+
+            // Discard trivial equivalence classes.
+            self.classes.retain(|class| class.len() > 1);
+            self.classes.sort();
+        }
+    }
+
+    /// Produce the equivalences present in both inputs.
+    pub fn union(&self, other: &Self) -> Self {
+        // TODO: seems like this could be extended, with similar concepts to localization:
+        //       We may removed non-shared constraints, but ones that remain could take over
+        //       and substitute in to retain more equivalences.
+
+        // For each pair of equivalence clasess, their intersection.
+        let mut equivalences = EquivalenceClasses {
+            classes: Vec::new(),
+        };
+        for class1 in self.classes.iter() {
+            for class2 in other.classes.iter() {
```

Comment:
Could you please add some limit here that just gives up if `self.classes.len() * other.classes.len()` is too large? (I've already [seen](https://materializeinc.slack.com/archives/C02PPB50ZHS/p1696337875755359?thread_ts=1696004419.256909&cid=C02PPB50ZHS) a join in a customer's MIR plan with 196 equivalences...)

Reply:
I'm conflicted on this, given the status of the codebase and the struggles with predictability of transforms. If we can figure out a principled way to do this that doesn't fight with idempotence, that doesn't make the outer fixpoint iteration brittle, etc., then yes! But if it's "don't do something quadratic in the input" then .. we have larger problems. :D

Reply:
Maybe we'll never run into the limit. How about you add a `soft_assert_or_log`? And then CI would fail if we run into the limit, and in production we'd get a Sentry error. And then if it never happens then everyone is happy, and if it happens then we can figure out the next steps (increasing the limit, optimizing the code, ignoring the problem if it doesn't look dangerous that the optimization was not applied in the specific situation, ...).

Reply:
Ah, I've only now seen your comment about using `HashSet`s. I'm not sure how to do it with `HashSet`s, but a possible approach would be to have a `HashMap` that maps from expressions to indexes in `classes` (relying on the fact that `minimize` makes it so that each expression can appear in only one class). And a similar thing should also work for my [other similar comment](https://github.com/MaterializeInc/materialize/pull/24155/files#r1503184823).